libavutil/hwcontext_dxcore.cpp

```diff
@@ -0,0 +1,119 @@
+/*
```

Comment:
ffmpeg is written in C, not C++. They do prefer to work in a C style with DX API. For example:
https://github.com/FFmpeg/FFmpeg/blob/4a80db5fc2b1b134550ffbcb8fd3b7ce2ad734b3/libavcodec/dxva2.c#L180
So, a call looks as:
```
hr = IDirectXVideoDecoderService_GetDecoderRenderTargets((IDirectXVideoDecoderService *)decoder_service, &guid, &target_count, &target_list);
```

Reply:
Unfortunately, DXCore does not have a C API available similar to the ones used in DXVA/D3D11VA. I separated the cpp code in this new file to avoid tainting the existing VAAPI C code. I see other .cpp files in the project, would this be an issue given that there's no pure C API available?

Reply:
I don't know. @galinart will probably be the best to comment since he worked with ffmpeg on Windows.

Reply:
@galinart kindly pinging on this :)

Reply:
@galinart @dvrogozh . To avoid any issues with C++ code given [the C code requirement in the ffmpeg developer guide](https://ffmpeg.org/developer.html#toc-C-language-features) I re-wrote the adapter selection logic following the [D3D11VA similar logic](https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/hwcontext_d3d11va.c#L581), please let me know if looks good to you. CC @tong1wu

Reply:
It is better to avoid cpp files but this is FFmpeg community decides if it is acceptable.